helix-dap/src/debuger_service.rs

```diff
@@ -0,0 +1,104 @@
+use crate::{Client, StackFrame};
+use std::collections::HashMap;
+
+/// DebuggerService is a struct that manages and owns multiple debugger clients
+/// This holds the responsibility of managing the lifecycle of each client
+/// plus showing the heirarcihical nature betweeen them
+pub struct DebuggerService {
+    /// The top-level debuggers being held by this service
+    debuggers: HashMap<usize, Client>,
+    /// The active debugger client
+    ///
+    /// TODO: You can have multiple active debuggers, so the concept of a single active debugger
+    /// may need to be changed
+    current_debugger_id: Option<usize>,
+    /// This is used to generate unique ids for each debugger
+    pub counter: usize,
+}
```

Comment:
See the `LanguageServerId` type and the use of `HopSlotMap` for language servers - that can be used instead of the hashmap+counter here and we probably don't need a new struct for this

Reply:
Thanks for looking.

I don’t know if HopSlotMap will work as you need to know the ID when creating the Client, so it’s too late by the time it’s added to the map. I believe this is different with language servers

Reply:
The LSP registry uses [`HopSlotMap::try_insert_with_key`](https://docs.rs/slotmap/latest/slotmap/struct.HopSlotMap.html#method.try_insert_with_key). It should be possible to use that for DAP as well to get the next ID and pass it to `Client::tcp`/`Client::process`.

Reply:
This has been resolved in the latest update